l1-contracts/contracts/bridge/L1ERC20Bridge.sol

```diff
@@ -6,122 +6,303 @@ import "@openzeppelin/contracts/token/ERC20/extensions/IERC20Metadata.sol";
+        address _owner
+    ) external payable reinitializer(2) {
+        _transferOwnership(_owner);
+        require(_l2TokenBeaconStandardAddress != address(0), "L1EB: tb  0");
+        require(_l2BridgeStandardAddress != address(0), "L1EB: bridge 0");
+        require(_owner != address(0), "L1EB: owner 0");
         // We are expecting to see the exact three bytecodes that are needed to initialize the bridge
         require(_factoryDeps.length == 3, "mk");
         // The caller miscalculated deploy transactions fees
-        require(msg.value == _deployBridgeImplementationFee + _deployBridgeProxyFee, "fee");
         l2TokenProxyBytecodeHash = L2ContractHelper.hashL2Bytecode(_factoryDeps[2]);
-        l2TokenBeacon = _l2TokenBeacon;
+        l2TokenBeaconStandardAddress = _l2TokenBeaconStandardAddress;
+        l2BridgeStandardAddress = _l2BridgeStandardAddress;
 
-        bytes32 l2BridgeImplementationBytecodeHash = L2ContractHelper.hashL2Bytecode(_factoryDeps[0]);
-        bytes32 l2BridgeProxyBytecodeHash = L2ContractHelper.hashL2Bytecode(_factoryDeps[1]);
+        // #if !EOA_GOVERNOR
+        uint32 size;
+        assembly {
+            size := extcodesize(_owner)
+        }
+        require(size > 0, "L1EB, owner cannot be EOA");
+        // #endif
 
-        // Deploy L2 bridge implementation contract
-        address bridgeImplementationAddr = BridgeInitializationHelper.requestDeployTransaction(
-            zkSync,
-            _deployBridgeImplementationFee,
-            l2BridgeImplementationBytecodeHash,
-            "", // Empty constructor data
-            _factoryDeps // All factory deps are needed for L2 bridge
-        );
+        factoryDepsHash = keccak256(abi.encode(_factoryDeps));
+    }
 
-        // Prepare the proxy constructor data
-        bytes memory l2BridgeProxyConstructorData;
+    function initializeChainGovernance(
+        uint256 _chainId,
+        address _l2BridgeAddress,
+        address _l2TokenBeaconAddress
+    ) external onlyOwner {
+        l2BridgeAddress[_chainId] = _l2BridgeAddress;
+        l2TokenBeaconAddress[_chainId] = _l2TokenBeaconAddress;
+    }
+
+    /// @dev Initializes a contract bridge for later use. Expected to be used in the proxy
+    /// @dev During initialization deploys L2 bridge counterpart as well as provides some factory deps for it
+    /// @param _factoryDeps A list of raw bytecodes that are needed for deployment of the L2 bridge
+    /// @notice _factoryDeps[0] == a raw bytecode of L2 bridge implementation
+    /// @notice _factoryDeps[1] == a raw bytecode of proxy that is used as L2 bridge
+    /// @notice _factoryDeps[2] == a raw bytecode of token proxy
+    /// @param _deployBridgeImplementationFee How much of the sent value should be allocated to deploying the L2 bridge
+    /// implementation
+    /// @param _deployBridgeProxyFee How much of the sent value should be allocated to deploying the L2 bridge proxy
+    function startErc20BridgeInitOnChain(
```

Comment:
This might be an issue. I think we should keep the function permissionless, and it should be overwritable. 

I see 3 options: 
1. the function can only be called again if the transaction failed
2. add a timegap, so there is time to finish initialization. 
3. store each deployment, based on salt or txHash, they don't overwrite each other. 

Probably 1 is best.

Reply:
The first one sounds as the simplest solution for me